src/core/SP_reader.sol

```diff
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: GNU AGPLv3
+pragma solidity ^0.8.20;
+
+import { OwnableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import { Attestation as SPAttestation } from "../models/Attestation.sol"; // Ignoring Offchain Attestations
+import { Attestation as VeraxAttestation, IVerax } from "src/interfaces/IVerax.sol";
+import { SP } from "src/core/SP.sol";
+/*
+    Found Information Above on Github:
+    - Verax's Attestation struct 
+    - Interface I created to call a certain function -- getAttestation()
+*/
+
+contract Reader is OwnableUpgradeable {
+    /// @custom:storage-location erc7201:ethsign.SP-Reader
+    struct ReaderStorage {
+        SP signProtocol;
+        IVerax veraxRegistry;
+    }
+
+    // keccak256(abi.encode(uint256(keccak256("ethsign.SP-Reader")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 private constant ReaderStorageLocation = 0xb209ab9212fdbbb86ea9ba187ed37f12ac1d5679ecdf9a605a02f993a67d7400;
+
+    error Invalid();
+    error Incompatible();
+    error DoesNotExist();
+
+    function _getReaderStorage() internal pure returns (ReaderStorage storage $) {
+        assembly {
+            $.slot := ReaderStorageLocation
+        }
+    }
+
+    constructor() {
+        if (block.chainid != 31_337) {
+            _disableInitializers();
+        }
+    }
+
+    function initialize(address _spAddress, address _veraxAddress) public initializer {
```

Comment:
Generally, parameter naming doesn't involve adding underscores before the names. You may have seen me add underscores after the names, that's purely to avoid variable shadowing. Since you are using ERC7201 namespacing for storage variables, that isn't a concern.